From d9e9591febcba86b98c4c49b9aab2d991bc9ff4c Mon Sep 17 00:00:00 2001
From: NikLeberg <niklaus.leuenb@gmail.com>
Date: Mon, 11 Aug 2025 13:19:08 +0200
Subject: [PATCH] jtag/drivers/jtag_dpi: fix wraparound bug in runtest

Commit 0847a4d7fb98 ("jtag/commands: Use 'unsigned int' data type")
introduced bug when changing loop variable from `int` to `unsigned int`.
Instead of getting negative and terminating the loop, the value wraps
around to `INT_MAX` and the loop never finishes.

Change-Id: I055025a1f8eb4abe50955607b3e89530dfd92af4
Signed-off-by: NikLeberg <niklaus.leuenb@gmail.com>
Fixes: 0847a4d7fb98 ("jtag/commands: Use 'unsigned int' data type")
---

diff --git a/src/jtag/drivers/jtag_dpi.c b/src/jtag/drivers/jtag_dpi.c
index d6418d3..35dd245 100644
--- a/src/jtag/drivers/jtag_dpi.c
+++ b/src/jtag/drivers/jtag_dpi.c
@@ -189,7 +189,7 @@
 		return ERROR_FAIL;
 	}
 	snprintf(buf, sizeof(buf), "ib %d\n", num_bits);
-	while (num_cycles > 0) {
+	for (unsigned int cycle = 0; cycle < num_cycles; cycle += num_bits + 6) {
 		ret = write_sock(buf, strlen(buf));
 		if (ret != ERROR_OK) {
 			LOG_ERROR("write_sock() fail, file %s, line %d",
@@ -208,8 +208,6 @@
 				__FILE__, __LINE__);
 			goto out;
 		}
-
-		num_cycles -= num_bits + 6;
 	}
 
 out:
